pyogrio/tests/test_path.py

```diff
@@ -0,0 +1,135 @@
+        # archives
+        ("zip://data.zip", "/vsizip/data.zip"),
+        ("tar://data.tar", "/vsitar/data.tar"),
+        ("gzip://data.gz", "/vsigzip/data.gz"),
+        ("tar://./my.tar!my.geojson", "/vsitar/./my.tar/my.geojson"),
+        (
+            "zip://home/data/shapefile.zip!layer.shp",
+            "/vsizip/home/data/shapefile.zip/layer.shp",
+        ),
+        # combined schemes
+        ("zip+s3://testing/shapefile.zip", "/vsizip/vsis3/testing/shapefile.zip"),
+        (
+            "zip+https://s3.amazonaws.com/testing/shapefile.zip",
+            "/vsizip/vsicurl/https://s3.amazonaws.com/testing/shapefile.zip",
+        ),
+    ],
+)
+def test_vsi_path(path, expected):
+    assert vsi_path(path) == expected
+
+
+def test_vsi_handling_read_functions(naturalearth_lowres_vsi):
+    # test that all different read entry points have the path handling
+    # (a zip:// path would otherwise fail)
+    path, _ = naturalearth_lowres_vsi
+    path = "zip://" + str(path)
+
+    result = pyogrio.raw.read(path)
+    assert len(result[1]) == 177
+
+    result = pyogrio.read_info(path)
+    assert result["features"] == 177
+
+    result = pyogrio.read_bounds(path)
+    assert len(result[0]) == 177
+
+    result = pyogrio.read_dataframe(path)
+    assert len(result) == 177
+
+
+def test_path_absolute(data_dir):
+    # pathlib path
+    path = data_dir / "naturalearth_lowres/naturalearth_lowres.shp"
+    df = pyogrio.read_dataframe(path)
+    len(df) == 177
+
+    # str path
+    df = pyogrio.read_dataframe(str(path))
+    len(df) == 177
+
+
+def test_path_relative(data_dir):
+    with change_cwd(data_dir):
+        df = pyogrio.read_dataframe("naturalearth_lowres/naturalearth_lowres.shp")
+    len(df) == 177
+
+
+def test_uri_local_file(data_dir):
+    uri = "file://" + str(data_dir / "naturalearth_lowres/naturalearth_lowres.shp")
+    df = pyogrio.read_dataframe(uri)
+    len(df) == 177
+
+
+def test_zip_path(naturalearth_lowres_vsi):
+    path, path_vsi = naturalearth_lowres_vsi
+    path_zip = "zip://" + str(path)
+
+    # absolute zip path
+    df = pyogrio.read_dataframe(path_zip)
+    assert len(df) == 177
+
+    # relative zip path
+    with change_cwd(path.parent):
+        df = pyogrio.read_dataframe("zip://" + path.name)
+    assert len(df) == 177
+
+    # absolute vsizip path
+    df = pyogrio.read_dataframe(path_vsi)
+    assert len(df) == 177
+
+
+@pytest.mark.network
+def test_url():
+    df = pyogrio.read_dataframe(
+        "https://raw.githubusercontent.com/geopandas/pyogrio/main/pyogrio/tests/fixtures/naturalearth_lowres/naturalearth_lowres.shp"
+    )
+    assert len(df) == 177
+
+
+@pytest.mark.network
+def test_url_with_zip():
+    df = pyogrio.read_dataframe(
+        "zip+https://s3.amazonaws.com/fiona-testing/coutwildrnp.zip"
+    )
+    assert len(df) == 67
+
+
+# @pytest.mark.network
+# def test_uri_s3()
+#     df = pyogrio.read_dataframe('zip+s3://fiona-testing/coutwildrnp.zip')
+#     assert len(df) == 67
```

Comment:
It would be nice to have an s3 test, but that also requires some set up. I also _think_ that fiona still has some more logic around setting up credentials for GDAL.

Reply:
(but this is of course not specific to this PR, the same thing is true for `/vsis3` which already can be done before this PR.

Reply:
For testing, I could copy the code from dask-geopandas (https://github.com/geopandas/dask-geopandas/blob/main/tests/io/conftest.py) for a fixture that sets up a local boto client to mimic s3. Although I actually don't know how that would work with gdal's vsis3 ..

Reply:
What is the issue here? This test passes locally. `fiona-testing` bucket seems to be public.

Reply:
Thanks!

I don't know if it matters much, but it seems you can also do that with a fixture for a specific test, although I suppose it wouldn't actually have any influence on the other tests